Add option to QgsFeatureRequest to request features within a certain distance of a reference geometry

nyalldawson opened the pull request:
E.g. this request will retrieve all features within 50 map units of the provided linestring:

    QgsFeatureRequest().setDistanceWithin(QgsGeometry.fromWkt('LineString(0 0, 10 0, 12 1)'), 50)

A new enum Qgis::SpatialFilterType has been added to reflect whether a request uses no spatial filter, a BoundingBox filter (via
setFilterRect), or the new DistanceWithin filter.

Distance within filters are treated like bounding box filters, in that they are independant of any attribute/id filters (such as
feature ids or expressions).

Provider feature iterators can potentially delegate the distance within search to the backend (e.g. postgres could use a ST_DWithin query) for optimal index use.

~~Currently handled only with the memory provider -- if the api and approach is deemed satisfactory then I'll implement for the other providers.~~

nyalldawson committed:
[api] Add option to QgsFeatureRequest to request features
within a certain distance of a reference geometry

E.g. this request will retrieve all features within 50 map units of the
provided linestring:

    QgsFeatureRequest().setDistanceWithin(QgsGeometry.fromWkt('LineString(0 0, 10 0, 12 1)'), 50)

A new enum Qgis::SpatialFilterType has been added to reflect whether
a request uses no spatial filter, a BoundingBox filter (via
setFilterRect), or the new DistanceWithin filter.

Distance within filters are treated like bounding box filters, in
that they are independant of any attribute/id filters (such as
feature ids or expressions).

nyalldawson committed:
Handle distance within feature requests in memory provider

nyalldawson committed:
Implement distance within for OGR provider

nyalldawson committed:
Update tests

nyalldawson committed:
Implement distance within for delimited text provider

nyalldawson committed:
Implement distance within search for vector layer iterator

nyalldawson committed:
Correct test string

nyalldawson committed:
Handle distance within in python provider

nyalldawson committed:
Implement distance within for gpx provider

nyalldawson committed:
Distance within for AFS provider

nyalldawson committed:
Working distance within filter for vector layer cache

nyalldawson committed:
Distance within for virtual provider

nyalldawson commented:
@3nids do you know why the actual oracle provider test isn't running on the Oracle workflow? Only the Oracle connection test runs...

3nids commented:
> @3nids do you know why the actual oracle provider test isn't running on the Oracle workflow? Only the Oracle connection test runs...

forwarding to @troopa81 as he's been setting this up

nyalldawson committed:
Implement distance within for WFS/oapif provider

nyalldawson committed:
Implement distance within for spatialite feature iterator

nyalldawson committed:
Untested implementation of distance within for db2 provider

This is impossible to test, and the provider is deprecated anyway...
at least it's a best guess effort.

nyalldawson committed:
Distance within for HANA provider

nyalldawson committed:
Distance within for SQL Server provider

nyalldawson committed:
Implement distance within for postgres iterator

Note: this could be optimised by converting automatically to
a ST_DWithin(...) clause!

nyalldawson committed:
Distance within for oracle provider

nyalldawson committed:
[postgres] Handoff distance within queries to backend to use ST_DWithin
wherever it's safe to do so

nyalldawson commented:
Thanks @nirvn!
